### ISSUE TITLE
Simplify the build process using git submodules

### ISSUE THREAD

**Issue by rwhitby** [kintex-chatter/xc7k325t-blinky-nextpnr#18]
I think the build process could be simplified by using git submodules:

1. sudo apt install libftdi1-dev libudev-dev git cmake build-essential tclsh clang tcl-dev libreadline-dev flex bison python3-dev libboost-all-dev libqt5-base-dev-tools libeigen3-dev python3 python3-pip python3-yaml pypy3 pkg-config libqt5opengl5-dev
2. clone/build/install yosys from https://github.com/YosysHQ/yosys or download a release from https://github.com/YosysHQ/oss-cad-suite-build/releases note: test have been performed with Yosys 0.13+28 (git sha1 bf85dfee5, gcc 10.2.1-6 -fPIC -Os)
3. git clone https://github.com/kintex-chatter/xc7k325t-blinky-nextpnr.git
> This would change to include --recurse-submodules
4. git clone --recurse-submodules https://github.com/kintex-chatter/nextpnr-xilinx.git
> This could be made a submodule of xc7k325t-blinky-nextpnr.git and then the branch and specific commit could be controlled.  that would remove this step.  The nextpnr-xilinx repository could also submodule the kintex7-specific prjxray-db repository.
5. cd nextpnr-xilinx
> This would change to cd xc7k325t-blinky-nextpnr/nextpnr-xilinx
6. git checkout xilinx-upstream
> This would no longer be required
7. mkdir build
8. pushd build
9. cmake -DARCH=xilinx -DCMAKE_INSTALL_PREFIX=~/opt/nextpnr ..
10. make -j2 && make install
11. popd
12. pushd xilinx/external
> This would no longer be required
13. rm -rf prjxray-db
> This would no longer be required
14. git clone -b k325 https://github.com/kintex-chatter/prjxray-db.git
> This would no longer be required
15. popd
> This would no longer be required
16. python3 xilinx/python/bbaexport.py --device xc7k325tffg676-1 --bba xilinx/xc7k325tffg676-1.bba
17. build/bbasm -l xilinx/xc7k325tffg676-1.bba xilinx/xc7k325tffg676-1.bin
18. mkdir -p ~/opt/nextpnr/xilinx-chipdb
19. ln -s $PWD/xilinx/external/prjxray-db ~/opt/nextpnr/
20. cp xilinx/xc7k325tffg676-1.bin ~/opt/nextpnr/xilinx-chipdb/

**Comment by rwhitby** [comment]
I will test this theory out, and if it works I'll submit a pull request.

**Comment by rwhitby** [comment]
First pull request (for adding prjxray.db into nextpnr-xilinx) is at https://github.com/kintex-chatter/nextpnr-xilinx/pull/1 which removes the need for steps 12 through 15.

Once this one is merged, I'll submit a second pull request to add nextpnr-xilinx to the current repo and remove steps 4 through 6.

**Comment by unbtorsten** [comment]
Thank you, for the suggestions @rwhitby! I agree that this will simplify development.

In the long run, our focus should be an update of upstream repositories, and a build procedure that clones directly from prjxray. This will only take an adjustment of the submodule. I opened https://github.com/kintex-chatter/nextpnr-xilinx/pull/2 to remind us of this, whenever upstream prjxray-db is updated correspondingly.

**Comment by unbtorsten** [comment]
@rwhitby: I have updated the README correspondingly. I'd say we're ready for the next step of adding [kintex-chatter/nextpnr-xilinx.git](https://github.com/kintex-chatter/nextpnr-xilinx.git) as submodule.

**Comment by rwhitby** [comment]
@unbtorsten Thanks for the branch fix.  Second step is coming soon.

**Comment by rwhitby** [comment]
Should we add https://github.com/kintex-chatter/prjxray.git as a submodule too? Then it will be in the known place for the Makefile to reference.

**Comment by rwhitby** [comment]
I also think that the top-level Makefile could automatically run most of these steps with an install target.

**Comment by unbtorsten** [comment]
> Should we add https://github.com/kintex-chatter/prjxray.git as a submodule too?
> 
> Then it will be in a known place for the Makefile to reference.

If you feel like automating this, I would opt for making this optional (if declined, tell the user to set path to prjxray manually). What do you think?

**Comment by unbtorsten** [comment]
> I also think that the top-level Makefile could automatically run most of these steps with a "setup" target.
> 
> [rwhitby#1](https://github.com/rwhitby/xc7k325t-blinky-nextpnr/pull/1) (this depends on #20 being merged first, so has not yet been submitted here to prevent confusion).

I added some comments in the above mentioned pull request for your fork to continue the discussion there.

**Comment by jrrk2** [comment]
I think yosys should also a a sub-repo since most distributions will not have a sufficiently up-to-date version available as a binary dependency

**Comment by rwhitby** [comment]
> I think yosys should also a a sub-repo since most distributions will not have a sufficiently up-to-date version available as a binary dependency

As no changes are made to yosys, I found it easiest just to install oss-cad-suite binary nightly and put it in the path.

**Comment by unbtorsten** [comment]
> I think yosys should also a a sub-repo [...]

Agreed. However, I notice that we are mixing `make setup` and `make blinky`, which is where things can get confusing.
I propose that we separate the build environment and tools that are the same for all packages
- yosys,
- prjxray,
- prjxray-db::k325,
- openFPGALoader
 
from the data that is specific to the packages
- nextpnr-xilinx?,
- constraints,
- blinky.v

Feel free to edit this comment to rearrange items between lists so that we can reach consensus.

**Comment by jrrk2** [comment]
If we want to be  portable to Mac-M1, I don’t think we can rely on oss-cad-suite.

**Comment by unbtorsten** [comment]
I think we should focus on and clone/build only those repositories that require changes in comparison to their respective upstream. Creating a universal FOSS build environment is the objective of Symbiflow et al. 

That said, I have updated the proposal for a scope of README and Makefile(s) stated in https://github.com/kintex-chatter/xc7k325t-blinky-nextpnr/issues/18#issuecomment-1046304962. My distribution, too, ships a version of yosys that is outdated with regard to our needs. Nevertheless, the required source and binaries are available from upstream. The same applies for prjxray. It is thus sufficient to rely on them being installed (in specific paths) as opposed to cloning them. The current revision of the README process reflects this and calls for installation of specific files in steps 0, 1, and 14.

**Comment by rwhitby** [comment]
I agree that anything upstream that is unchanged, should just be installed/used from upstream, and pull requests sent as soon as possible to incorporate any changes to upstream (as I have done already for openFPGALoader).

**Comment by rwhitby** [comment]
Pull request https://github.com/kintex-chatter/xc7k325t-blinky-nextpnr/pull/25 covers the second part of my proposal.
